### ISSUE TITLE
Unable to fetch translations with the new API?

### ISSUE THREAD

**Issue by cedx** [Smartling/api-sdk-php#27]
I was using a previous API code (when there was no dependency on Guzzle client, probably from june 2015) that worked fine. Here is a functional sample:

```php
$smartling = new \SmartlingAPI(null, $config['apiKey'], $config['projectId'], \SmartlingApi::PRODUCTION_MODE);
$response = $smartling->downloadFile("/AvePublishing-Store/$file", 'fr-FR');
```

The `$response` variable contained a JSON string containing all my translations.
My problem is that the current API code doesn't work anymore. I've updated my code to the new API:

```php
use GuzzleHttp\Client;
use Smartling\SmartlingAPI;

$client = new Client(['base_uri' => SmartlingApi::PRODUCTION_URL]);
$smartling = new SmartlingAPI($client->getConfig('base_uri'), $config['apiKey'], $config['projectId'], $client, SmartlingApi::PRODUCTION_MODE);
$response = $smartling->downloadFile("/AvePublishing-Store/$file", 'fr-FR')->getBody();
```

The `$response` variable contains exactly the same JSON content as before, but the new API raises this error everytime :

```
Notice: Undefined index: response in /repo/avepublishing-store/vendor/smartling/api-sdk-php/src/SmartlingAPI.php on line 130

Warning: implode(): Invalid arguments passed in /repo/avepublishing-store/vendor/smartling/api-sdk-php/src/SmartlingAPI.php on line 130

Fatal error: Uncaught exception 'Smartling\SmartlingApiException' in /repo/avepublishing-store/vendor/smartling/api-sdk-php/src/SmartlingAPI.php:130

Stack trace:
#0 /repo/avepublishing-store/vendor/smartling/api-sdk-php/src/SmartlingAPI.php(221): Smartling\SmartlingApi->sendRequest('file/get', Array, 'GET')
#1 /repo/avepublishing-store/lib/server/Application.php(139): Smartling\SmartlingApi->downloadFile('/AvePublishing-...', 'fr-FR')
#2 /repo/avepublishing-store/lib/server/Application.php(194): aquafadas\Application->loadMessageSources()
#3 [internal function]: aquafadas\Application->onBeforeRequest(Object(yii\base\Event))
#4 /repo/avepublishing-store/vendor/yiisoft/yii2/base/Component.php(541): call_user_func(Array, Object(yii\base\Event))
#5 /repo/avepublishing-store/vendor/yiisoft/yii2/base/Application.php(372): yii\base\Component->trigger('beforeRequest')
#6 /repo/avepublishing-store/web/index.php(25): yii\base\Application->run()
#7 {main} thrown in /repo/avepublishing-store/vendor/smartling/api-sdk-php/src/SmartlingAPI.php on line 130
```

The original translation file is a plain JSON file, without any Smartling directive.

**Comment by dimitrystd** [comment]
@cedx just merged fix #28

**Comment by cedx** [comment]
No luck :( The new code did not fix my issue:

```php
<br />
<b>Notice</b>:  Undefined index: response in <b>/repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php</b> on line <b>130</b><br />
<br />
<b>Warning</b>:  implode(): Invalid arguments passed in <b>/repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php</b> on line <b>130</b><br />
<br />
<b>Fatal error</b>:  Uncaught exception 'Smartling\SmartlingApiException' in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php:130
Stack trace:
#0 /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php(221): Smartling\SmartlingApi-&gt;sendRequest('file/get', Array, 'GET')
#1 /repo/aquafadas.store/lib/server/Application.php(143): Smartling\SmartlingApi-&gt;downloadFile('/AvePublishing-...', 'fr-FR')
#2 /repo/aquafadas.store/lib/server/Application.php(198): aquafadas\Application-&gt;loadMessageSources()
#3 [internal function]: aquafadas\Application-&gt;onBeforeRequest(Object(yii\base\Event))
#4 /repo/aquafadas.store/vendor/yiisoft/yii2/base/Component.php(541): call_user_func(Array, Object(yii\base\Event))
#5 /repo/aquafadas.store/vendor/yiisoft/yii2/base/Application.php(372): yii\base\Component-&gt;trigger('beforeRequest')
#6 /repo/aquafadas.store/web/index.php(25): yii\base\Application-&gt;run()
#7 {main}
  thrown in <b>/repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php</b> on line <b>130</b><br />
```

When I try to log the values returned by the Guzzle client, I get (see file `src/SmartlingApi.php`, the `sendRequest()` method on line 125):

```php
var_dump($guzzle_response);
// object(GuzzleHttp\Psr7\Response) ...

var_dump($response);
// array(476) {
//  ["asset.allIssues"]=>
//  string(117) "Toutes les publications sont compatibles avec l’AppStore d’Apple, Google Play, Amazon AppStore et le web (HTML5)."
//  ["asset.allPlatforms"] ...
```

The response contains exactly the same JSON-decoded content than my source translation file:

```json
{
  "asset.allIssues": "All issues are ready for the Apple AppStore, Google Play, Amazon Appstore and the Web (HTML5).",
  "asset.allPlatforms": "Today all platforms (Apple AppStore, Google Play, Amazon Appstore and the Web) are included for publication licenses and application packs.",
...
...
}
```

So for me, this issue is "expected": the response data doesn't contains any `response` field at all.
What's wrong? Is it my code? My source translation file?

**Comment by cedx** [comment]
No luck :( The new code did not fix my issue:

```php
Notice: Undefined index: response in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php on line 135

Notice: Undefined index: response in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php on line 136

Notice: Undefined index: response in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php on line 137

Warning: implode(): Invalid arguments passed in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php on line 137

Fatal error: Uncaught exception 'Smartling\SmartlingApiException' in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php:137
Stack trace:
#0 /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php(228): Smartling\SmartlingApi->sendRequest('file/get', Array, 'GET')
#1 /repo/aquafadas.store/lib/server/Application.php(143): Smartling\SmartlingApi->downloadFile('/AvePublishing-...', 'fr-FR')
#2 /repo/aquafadas.store/lib/server/Application.php(198): aquafadas\Application->loadMessageSources()
#3 [internal function]: aquafadas\Application->onBeforeRequest(Object(yii\base\Event))
#4 /repo/aquafadas.store/vendor/yiisoft/yii2/base/Component.php(541): call_user_func(Array, Object(yii\base\Event))
#5 /repo/aquafadas.store/vendor/yiisoft/yii2/base/Application.php(372): yii\base\Component->trigger('beforeRequest')
#6 /repo/aquafadas.store/web/index.php(25): yii\base\Application->run() #7 {main} thrown in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php on line 137
```

When I try to log the values returned by the `SmartlingApi` client, I get (see file `src/SmartlingApi.php`, the `sendRequest()` method on line 125):

```php
var_dump($guzzle_response);
// object(GuzzleHttp\Psr7\Response) ...

var_dump($response_body);
// array(476) {
//  ["asset.allIssues"]=>
//  string(117) "Toutes les publications sont compatibles avec l’AppStore d’Apple, Google Play, Amazon AppStore et le web (HTML5)."
//  ["asset.allPlatforms"] ...
```

The response body already contains the same JSON-decoded content than my source translation file:

```json
{
  "asset.allIssues": "All issues are ready for the Apple AppStore, Google Play, Amazon Appstore and the Web (HTML5).",
  "asset.allPlatforms": "Today all platforms (Apple AppStore, Google Play, Amazon Appstore and the Web) are included for publication licenses and application packs.",
  ...
}
```

So for me, this issue is "expected": the response data doesn't contains any `response` field at all.
What's wrong? Is it my code? My source translation file?

**Comment by M1r1k** [comment]
Yeh, problem is that we were expecting XML only, let me fix this.

**Comment by dimitrystd** [comment]
Merged #29 which should fix issue

**Comment by cedx** [comment]
Still an issue, but this time it is related to the authentication. My new code:

```
$client=new Client(['base_uri'=>SmartlingApi::DEFAULT_SERVICE_URL]);
$smartling=new SmartlingApi($config['apiKey'], $config['projectId'], $client, $client->getConfig('base_uri'));
$response = $smartling->downloadFile("/AvePublishing-Store/$file", 'fr-FR');
```

The new error:

```
<br />
<b>Fatal error</b>:  Uncaught exception 'Smartling\SmartlingApiException' with message 'Bad response format from Smartling' in /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php:126
Stack trace:
#0 /repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php(232): Smartling\SmartlingApi-&gt;sendRequest('file/get', Array, 'GET')
#1 /repo/aquafadas.store/lib/server/Application.php(143): Smartling\SmartlingApi-&gt;downloadFile('/AvePublishing-...', 'fr-FR')
#2 /repo/aquafadas.store/lib/server/Application.php(203): aquafadas\Application-&gt;loadMessageSources()
#3 [internal function]: aquafadas\Application-&gt;onBeforeRequest(Object(yii\base\Event))
#4 /repo/aquafadas.store/vendor/yiisoft/yii2/base/Component.php(541): call_user_func(Array, Object(yii\base\Event))
#5 /repo/aquafadas.store/vendor/yiisoft/yii2/base/Application.php(372): yii\base\Component-&gt;trigger('beforeRequest')
#6 /repo/aquafadas.store/web/index.php(27): yii\base\Application-&gt;run()
#7 {main}
  thrown in <b>/repo/aquafadas.store/vendor/smartling/api-sdk-php/src/SmartlingApi.php</b> on line <b>126</b><br />
```

When I log the `$error_response` content, I get (line 123):

```
{
  "response": {
    "code": "AUTHENTICATION_ERROR",
    "requestId": "eb252d55-9893-4683-ad49-7662c6875dc8",
    "errors": [
      {
        "message": "Invalid token",
        "key": "invalid_token",
        "details": {}
      }
    ]
  }
}
```

**Comment by dimitrystd** [comment]
Fixed in #31. It was bug in code and tests :(

**Comment by cedx** [comment]
Sorry, very busy these last weeks. Tested today the new code with success :) Thanks.
Now, I'm using the `dev-master#06fe9ab8bd56bf642a23af7087663de23ca52044` version in my `composer.json` file.
